libs/providers/config-cat-web/src/lib/config-cat-web-provider.ts

```diff
@@ -0,0 +1,165 @@
+import {
+  EvaluationContext,
+  FlagNotFoundError,
+  JsonValue,
+  OpenFeatureEventEmitter,
+  Paradigm,
+  ParseError,
+  Provider,
+  ProviderEvents,
+  ProviderNotReadyError,
+  ResolutionDetails,
+} from '@openfeature/web-sdk';
+import { TypeMismatchError } from '@openfeature/core';
+import {
+  isType,
+  PrimitiveType,
+  PrimitiveTypeName,
+  toResolutionDetails,
+  transformContext,
+} from '@openfeature/config-cat-core';
+import {
+  getClient,
+  IConfig,
+  IConfigCatClient,
+  IConfigCatClientSnapshot,
+  OptionsForPollingMode,
+  PollingMode,
+} from 'configcat-js-ssr';
+
+export class ConfigCatWebProvider implements Provider {
+  public readonly events = new OpenFeatureEventEmitter();
+  private readonly _sdkKey: string;
+  private readonly _configCatOptions: OptionsForPollingMode<PollingMode.AutoPoll>;
+  private _client?: IConfigCatClient;
+  private _clientSnapshot?: IConfigCatClientSnapshot;
+
+  public runsOn: Paradigm = 'client';
+
+  public metadata = {
+    name: ConfigCatWebProvider.name,
+  };
+
+  constructor(sdkKey: string, options: OptionsForPollingMode<PollingMode.AutoPoll> = {}) {
+    this._sdkKey = sdkKey;
+    this._configCatOptions = options;
+  }
+
+  public static create(sdkKey: string, options: OptionsForPollingMode<PollingMode.AutoPoll> = {}) {
+    return new ConfigCatWebProvider(sdkKey, options);
+  }
+
+  public async initialize(): Promise<void> {
+    const options = this._configCatOptions;
+    const oldSetupHooks = options.setupHooks;
+
+    options.setupHooks = (hooks) => {
+      oldSetupHooks?.(hooks);
+
+      hooks.on('clientReady', () => {
+        this.events.emit(ProviderEvents.Ready);
+      });
+
+      hooks.on('configChanged', (projectConfig: IConfig | undefined) => {
+        this.events.emit(ProviderEvents.ConfigurationChanged, {
+          flagsChanged: projectConfig ? Object.keys(projectConfig.settings) : undefined,
+        });
+
+        if (this._client) {
+          this._clientSnapshot = this._client.snapshot();
+        }
+      });
+
+      hooks.on('clientError', (message: string, error) => {
+        this.events.emit(ProviderEvents.Error, {
+          message: message,
+          metadata: error,
+        });
+      });
+    };
+
+    const client = getClient(this._sdkKey, PollingMode.AutoPoll, this._configCatOptions);
+    await client.waitForReady();
+
+    this._client = client;
+    this._clientSnapshot = client.snapshot();
```

Comment:
I advise against storing the snapshot as that would break advanced setups using a [distributed/shared cache](https://configcat.com/docs/advanced/caching/#shared-cache). (The `configChanged` event only fires when a new version of the config is fetched from the remote server, but not when an other actor puts a new version in the shared cache.)

Creating a snapshot is a cheap operation, so we can do it on demand.

Reply:
Oh okay, I did not consider this cheap but it makes sense.
Changed it @adams85.